### PR TITLE
84 slack notifications

### LINKED ISSUE #84: Slack Notifications
**Is your feature request related to a problem? Please describe.**
UserStory

**Describe the solution you'd like**
Slack notification.

### PR DESCRIPTION
**Pull request by nathanbillis**
This adds the ability for slack notifications, to close #84. There is currently a demo slackbot setup. This setup is designed for a single slack server, a future development could allow instances to install the bot on their own slack.

Messages are currently only the subject headline, mainly because the notify function only deals with email templates.

## Example Notification
Screen Shot 2021-06-03 at 14 47 25

### CODE REVIEW

File: html/admin/api/notifications/slack/slack.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    $token = $CONFIG['SLACK_BOT_OAUTH'];
+
+    // TODO add more detailed templates for slack notifications, will involve looking at the notify function because atm it only deals with email templates
+    if ($token != null) {
+        $data = array(
+            "channel" => $channel_id,
+            'type' => 'plain_text',
+            'text' => $subject
+        );
```
Comment (nathanbillis) [inline html/admin/api/notifications/slack/slack.php:21]: There's a commit showing how to make more complicated messages, it works but the preview breaks, so until we add templates this is the most elegant option

File: html/common/config.php [hunk 2]
```diff
@@ -54,8 +54,15 @@
                 'secret' => getenv('bCMS__OAUTH__GOOGLESECRET')
             ],
             'scope' => 'https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email'
+        ],
+        'SLACK' => [
+            'keys' => [
```
Comment (Jbithell) [inline html/common/config.php:59]: Where are these keys used?

**Comment by Jbithell** [comment]
Woo this is wonderful thanks so much @nathanbillis. What scopes and callback URLs does this require - could you add that to the readme?

File: html/common/config.php (hunk 2 above)
Comment (nathanbillis) [reply html/common/config.php:59]: Line 50 in linkoauth.php - basically the same as google, they are taken from the Adam-rms bot settings on slack

**Comment by Jbithell** [comment]
I think this is going to have to be parked until we can figure out how to get it to support more than one slack instance

**Review by Jbithell — CHANGES_REQUESTED** [review]
Needs to support multiple slack instances

**Comment by nathanbillis** [comment]
@Jbithell I'll have a look into getting multiple slack instances up and running :)

**Comment by nathanbillis** [comment]
@Jbithell - Just unassigned myself from this because I won't have time to pick it up. Workflows/webhooks etc could be an idea. Might eventually have time but just not at the moment 😞

**Comment by Jbithell** [comment]
I'm going to go ahead and close this for now as some housekeeping. Significant work would be required to merge main back into this PR.

**Pull request closed without merging**